src/gc/impl/conservative/gc.d

```diff
@@ -165,6 +165,8 @@ class ConservativeGC : GC
     this()
     {
         //config is assumed to have already been initialized
+        if (config.fork && config.parallel > 0)
+            throw new Error("No reason in mixing threads and fork");
```

Comment:
@rainers this still results in a deadlock if the user does not specify 'fork:1 parallel:0'

Where do I throw this error avoiding the recursive lock?

Reply:
You can use a statically allocated error:
```
__gshared Error err = new Error("No reason in mixing threads and fork");
throw err;
```

Reply:
we have `throw staticError!Error` for that.

Reply:
staticError creates a deadlock in any case.
As you can see my `onConfigurationError` called during gc_init calls the traceHandler that needs the GC again. We can work around this issue with @nogc exceptions I believe or by aborting.

```
(gdb) bt 
#0  0x00007ffff7bc7c31 in __GI___nanosleep (requested_time=0x7fffffffdb00, remaining=0x7fffffffdb10) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  0x00005555557da1f4 in _D4core6thread6Thread5sleepFNbNiSQBf4time8DurationZv (val=...) at src/core/thread.d:1311
#2  0x000055555586653e in core.internal.spinlock.SpinLock.yield(ulong) shared (this=..., k=1015) at src/core/internal/spinlock.d:57
#3  0x00005555558664ba in core.internal.spinlock.SpinLock.lock() shared (this=...) at src/core/internal/spinlock.d:39
#4  0x000055555586187c in gc_init () at src/gc/proxy.d:55
#5  0x000055555584d121 in gc_init_nothrow () at src/gc/proxy.d:90
#6  0x000055555581e028 in _D2gc4impl5protoQo7ProtoGC6mallocMFNbmkxC8TypeInfoZPv (this=0x555555b3c5e0 <internal>, 
    ti=0x555555b49a30 <ClassInfo for core.runtime.defaultTraceHandler(void*).DefaultTraceInfo>, bits=0, size=1064) at src/gc/impl/proto/gc.d:104
#7  0x00005555557dcf7b in gc_malloc (sz=1064, ba=0, ti=0x555555b49a30 <ClassInfo for core.runtime.defaultTraceHandler(void*).DefaultTraceInfo>) at src/gc/proxy.d:165
#8  0x00005555557d92fb in core.memory.GC.malloc(ulong, uint, const(TypeInfo)) (ti=0x555555b49a30 <ClassInfo for core.runtime.defaultTraceHandler(void*).DefaultTraceInfo>, 
    ba=0, sz=1064) at src/core/memory.d:399
#9  0x00005555557dfbca in _d_newclass (ci=0x555555b49a30 <ClassInfo for core.runtime.defaultTraceHandler(void*).DefaultTraceInfo>) at src/rt/lifetime.d:94
#10 0x000055555584b0d9 in core.runtime.defaultTraceHandler(void*) (ptr=0x0) at src/core/runtime.d:983
#11 0x000055555582022f in _d_traceContext (ptr=0x0) at src/rt/dmain2.d:295
#12 0x000055555581ffeb in _d_createTrace (o=0x7ffff7fec410, context=0x0) at src/rt/deh.d:24
#13 0x00005555557de91d in _d_throwdwarf (o=0x7ffff7fec410) at src/rt/dwarfeh.d:219
#14 0x0000555555874e97 in onConfigurationError (msg=..., file=..., line=170) at src/core/exception.d:625
#15 0x00005555558679f0 in _D2gc4impl12conservativeQw14ConservativeGC6__ctorMFZCQBzQBzQBxQCiQBn (this=0x555555b54910) at src/gc/impl/conservative/gc.d:170
#16 0x0000555555867783 in _D2gc4impl12conservativeQw10initializeFZC4coreQBs11gcinterface2GC () at src/gc/impl/conservative/gc.d:133
#17 0x0000555555865546 in _D4core2gc8registry16createGCInstanceFAyaZCQBpQBn11gcinterface2GC (name=...) at src/core/gc/registry.d:64
#18 0x00005555558618c4 in gc_init () at src/gc/proxy.d:61
#19 0x000055555584d121 in gc_init_nothrow () at src/gc/proxy.d:90
#20 0x000055555581e028 in _D2gc4impl5protoQo7ProtoGC6mallocMFNbmkxC8TypeInfoZPv (this=0x555555b3c5e0 <internal>, ti=0x0, bits=0, size=48) at src/gc/impl/proto/gc.d:104
#21 0x00005555557dcf7b in gc_malloc (sz=48, ba=0, ti=0x0) at src/gc/proxy.d:165
#22 0x00005555557d92fb in core.memory.GC.malloc(ulong, uint, const(TypeInfo)) (ti=0x0, ba=0, sz=48) at src/core/memory.d:399
#23 0x00005555557dfb2d in _d_allocmemory (sz=48) at src/rt/lifetime.d:60
#24 0x000055555573a310 in D main (args=...) at dustmite.d:99
```

Reply:
> As you can see my `onConfigurationError` called during gc_init calls the traceHandler that needs the GC again.

That totally defeats the purpose of staticError ;-( It is only tricking the compiler...

Reply:
> That totally defeats the purpose of staticError

...and of "`@nogc` exceptions". In the case of the configuration check, you can just abort similar to https://github.com/dlang/druntime/blob/master/src/gc/proxy.d#L67